intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/verifiers/PluginUntilBuildVerifier.kt

```diff
@@ -0,0 +1,123 @@
+  }
+
+  private fun verifyBaseline(baseline: Int,
+                             untilBuildValue: String,
+                             untilBuild: IdeVersion?,
+                             descriptorPath: String,
+                             problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    if (baseline >= FIRST_YEARLY_BASED_RELEASE_NUMBER_YEAR) {
+      registerProblem(InvalidUntilBuild(descriptorPath, untilBuildValue, untilBuild))
+    } else if (baseline >= 999) {
+      registerProblem(InvalidUntilBuildWithMagicNumber(descriptorPath, untilBuildValue, baseline.toString()))
+    } else if (baseline >= SUSPICIOUS_BASELINE_LOWER_BOUND) {
+      registerProblem(SuspiciousUntilBuild(untilBuildValue))
+    } else {
+      verifyInThreeReleasesPerYear(untilBuildValue, baselineVersion = baseline, problemRegistrar)
+    }
+  }
+
+  private fun verifyMagicNumber(untilBuildValue: String,
+                                descriptorPath: String,
+                                problemRegistrar: ProblemRegistrar): ValidationResult = with(problemRegistrar) {
+    val untilBuild = untilBuildValue.toIntOrNull() ?: return ValidationResult.NOT_APPLICABLE
+    return if (untilBuild >= 999) {
+      registerProblem(InvalidUntilBuildWithMagicNumber(descriptorPath, untilBuildValue, untilBuildValue))
+      INVALID
+    } else {
+      VALID
+    }
+  }
+
+  private enum class ValidationResult {
+    VALID, INVALID, NOT_APPLICABLE
+  }
+
+  private fun verifySingleComponentUntilBuild(untilBuild: String,
+                                              descriptorPath: String,
+                                              problemRegistrar: ProblemRegistrar) {
+    try {
+      val untilBuildNumber = untilBuild.toInt()
+      verifyBaseline(untilBuildNumber, untilBuild, untilBuild = null, descriptorPath, problemRegistrar)
+    } catch (e: NumberFormatException) {
+      problemRegistrar.registerProblem(InvalidUntilBuild(descriptorPath, untilBuild))
+    }
+  }
+
+  private fun verifyInThreeReleasesPerYear(untilBuildValue: String,
+                                           baselineVersion: Int,
+                                           problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    if (baselineVersion >= FIRST_YEARLY_BASED_RELEASE_NUMBER_BASELINE) {
+      val lastDigit = baselineVersion % 10
+      val releaseVersion = baselineVersion / 10
+      if (lastDigit !in 1..3) {
+        registerProblem(NonexistentReleaseInUntilBuild(untilBuildValue, "20$releaseVersion.$lastDigit"))
+      }
+    }
+  }
+
+  private fun isSpecialSingleComponent(untilBuild: String): Boolean {
+    return BUILD_NUMBER == untilBuild || SNAPSHOT == untilBuild
+  }
+
+  private fun isJustASingleComponent(untilBuild: String): Boolean {
+    if (!untilBuild.contains('.') && untilBuild.isNotBlank()) {
+      return true
+    }
+    val components = untilBuild.split('.').size
+    return components == 1
```

Comment:
Hm, probably, it shouldn't be possible to upload the plugin with until like '241.'
I don't see such plugins in our database, but am not sure that we have a check for it...

Reply:
Ah, we'll get a null in `createIdeVersionIfValid`, since it's not valid.

Reply:
`241.` (dot-suffix) triggers an `InvalidUntilBuild`. 

Added a unit test to cover that.